src/refda/exec.c

```diff
@@ -198,32 +360,77 @@ void *refda_exec_worker(void *arg)
     {
         refda_msgdata_t item;
 
-        sem_wait(&(agent->execs_sem));
-        if (!refda_msgdata_queue_pop(&item, agent->execs))
+        refda_timeline_it_t tl_it;
+        refda_timeline_it(tl_it, agent->exec_timeline);
+        if (!refda_timeline_end_p(tl_it))
         {
-            // shouldn't happen
-            CACE_LOG_WARNING("failed to pop from execs queue");
-            continue;
+            const refda_timeline_event_t *next = refda_timeline_cref(tl_it);
+            if (cace_log_is_enabled_for(LOG_DEBUG))
+            {
+                struct timespec nowtime;
+                clock_gettime(CLOCK_REALTIME, &nowtime);
+
+                struct timespec diff = timespec_sub(next->ts, nowtime);
+
+                string_t buf;
+                string_init(buf);
+                timeperiod_encode(buf, &diff);
+                CACE_LOG_DEBUG("waiting for exec event or %s", string_get_cstr(buf));
+                string_clear(buf);
+            }
+
+            sem_timedwait(&(agent->execs_sem), &(next->ts));
+
+            struct timespec nowtime;
+            clock_gettime(CLOCK_REALTIME, &nowtime);
+
+            // execute appropriate callbacks now
+            while (next && timespec_le(next->ts, nowtime))
+            {
+                CACE_LOG_DEBUG("running deferred callback");
+                bool finished = (next->callback)(next->item);
+                if (finished)
```

Comment:
**Question**: Trying to confirm behavior here. So, if the callback has failed to finish then it will be removed from the execution timeline without the callback being called. If the application isn't done with the item, it has a chance to push the item back on the timeline within `callback`. This cycle could continue indefinitely if necessary.

Is that correct or am I misunderstanding how this API works?

Reply:
Somewhat; the timeline is really for internal processing of `ietf-dtnma-agent` controls. Any application can defer finish with a completion callback and will need to "wake" the exec worker with some non-execution sentinel item in the execs queue. This needs to be documented with an example added to unit tests.

Reply:
This need is addressed by #82 to allow closing this PR now.